server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -341,909 +178,24 @@ public class Node implements Closeable {
      * @param environment         the initial environment for this node, which will be added to by plugins
      */
     public Node(Environment environment) {
-        this(environment, PluginsService.getPluginsServiceCtor(environment), true);
+        this(NodeConstructor.construct(environment, new NodeServiceProvider(), true));
```

Comment:
My only concern with that is that there would be a lot of constructor parameters - 11 in total. The previous uses of builders I've used pass the builder object itself into the constructor so that we don't need to change the constructor signature if we add/remove parameters, and there's no chance of mixing up parameters with the same type.

The type hierarchy I was going for is that NodeConstructor is subsidiary to Node - it's an implementation detail of constructing a Node. Then it doesn't matter that there's a two-way relationship between them, as it's a detail of Node implementation.

Reply:
++
The concern about lots of parameters is valid, but it will be called in one place, and completely "private" (only the "NodeConstructor" or whatever we end up calling it will know it) reducing the probability of error a lot.
Also, I do think that breaking when we need to add/remove parameters (need to change the constructor signature) is actually a good thing - it will break in one place and force you think at sensible defaults or how to handle the situation, check invariants and preconditions etc.

Reply:
I've used the same approach as in [here](https://blogs.oracle.com/javamagazine/post/exploring-joshua-blochs-builder-design-pattern-in-java), except used `Construction` rather than `Builder` to show that it's a lot more involved than simply setting some fields & doing some validation. `NodeConstruction` would be an inner class of `Node`, but its far too big. That's where it logically belongs though.

Reply:
Like I said in my other comment, I think that calling it something other than `Builder` is a good idea - we are not taking from the user (as in the developer that wants to create a `Node`) simple bits of information to fill in fields needed to build an object. Rather, we are constructing a complex object through a complex process. 
And as long as I would like it to be an inner class as in the article (thanks for linking it), I agree it is too big.